HPCC-34509 Provide IUserDescriptor to FileReadPropertiesUpdater

shamser opened the pull request:
IUserDescriptor is provided to FileSprayer so file properties can be queried with the correct user permissions.

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [x] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

jakesmith reviewed (commented):
@shamser - I've only glanced at the changes, but I see some IDistributedFileSystem methods have been updated that deal with IFileDescriptor's only, e.g. "transfer" and one fof the forms of "replicate", but is the IUserDescriptor needed as widely?
updateSourceProperties() (and the IFileReadPropertiesUpdater) is only used if distributedSource (IDistributedFile) is set.

Also, the default nullptr's (IUserDescriptor *userdesc=nullptr), in the cases where dealing with a IDistributedFile that might involve an update of owners, is it ever right that no user is passed ? I suspect the methods that can take a IUserDescriptor should not allow none to be passed in.

jakesmith reviewed (changes requested):
@shamser - as discussed. Use IUserDesc via distribute file (will need to expose via new queryUser method).

shamser committed:
HPCC-34509 Provide IUserDescriptor to FileReadPropertiesUpdater

This change provides the IUserDescriptor to the FileReadPropertiesUpdater so that
the it can lookup files using the correct permissions.

Signed-off-by: Shamser Ahmed <shamser.ahmed@lexisnexisrisk.com>

jakesmith commented on dali/base/dadfs.cpp:
```diff
@@ -3509,6 +3509,11 @@ class CDistributedFileBase : implements INTERFACE, public CInterface
             queryAttributes().setPropInt("@expireDays", expireDays);
     }
+
+    virtual IUserDescriptor *queryUser()
```
trivial: missing override

jakesmith commented on dali/base/dadfs.hpp:
```diff
@@ -447,6 +447,8 @@ interface IDistributedFile: extends IInterface
     virtual bool getReadCost(cost_type &cost, bool calculateIfMissing=false) const = 0;
     virtual bool getWriteCost(cost_type &cost, bool calculateIfMissing=false) const = 0;
+
+    virtual IUserDescriptor *queryUser() = 0;
```
should be a const method.

jakesmith reviewed (commented):
@shamser - looks good, a couple of minor comments.

jakesmith reviewed (approved):
@shamser - looks good. Please squash.

ghalliday commented on esp/clients/ws_dfsclient/ws_dfsclient.cpp:
```diff
@@ -330,6 +330,10 @@ class CServiceDistributedFileBase : public CSimpleInterfaceOf<INTERFACE>
     }
+    virtual IUserDescriptor *queryUser() const override
+    {
+        UNIMPLEMENTED_X("CServiceDistributedFileBase::queryUser");
```
This should pass through userDesc from the 874,19:     createLegacyDFSFile/lookupLegacyDFSFile functions and return it.

ghalliday reviewed (commented):
@shamser The second implementation should return a real user, not thrown an exception.

ghalliday reviewed (commented):
I don't think that function name is right and...
@jakesmith IDFSFile::queryUserDescriptor links the result.  That looks wrong.

ghalliday commented on esp/clients/ws_dfsclient/ws_dfsclient.cpp:
```diff
@@ -330,6 +330,10 @@ class CServiceDistributedFileBase : public CSimpleInterfaceOf<INTERFACE>
     }
+    virtual IUserDescriptor *queryUser() const override
+    {
+        return legacyDFSFile->queryUser();
```
I may not be following the code correctly, but I don't think this has the right information to be able to implement this.
legacyDFSFile is a member of the base.  It only seems to be intialised by CServiceSuperDistributedFile, by a call that does not provide the user.
I think it could be accessed from dfsFile->queryUserDescriptor(), but that returns a linked pointer.
